Libraries/Components/FocusableWindows/FocusableWindows.windows.js

```diff
@@ -0,0 +1,270 @@
+ *                  style = {styleSet.childStyle}
+ *              />
+ *            </FocusableWindows>
+ *         );
+ *      }
+ *
+ *      // important for animations
+ *      setNativeProps (props) {
+ *          var splitProps = FocusableWindows.splitNativeProps(props);
+ *          if (splitProps.focusableProps !== undefined) {
+ *             this._focusable.setNativeProps(splitProps.focusableProps)
+ *          }
+ *          if (splitProps.childProps !== undefined) {
+ *             this._child.setNativeProps(splitProps.childProps);
+ *          }
+ *      }
+ *   }
+ *
+ * @providesModule FocusableWindows
+ */
+'use strict';
+
+var PropTypes = require('prop-types');
+var React = require('React');
+var ReactNative = require('ReactNative');
+var ViewPropTypes = require('ViewPropTypes');
+var requireNativeComponent = require('requireNativeComponent');
+var UIManager = require('UIManager');
+const flattenStyle = require('flattenStyle');
+
+class FocusableWindows extends React.Component {
+
+  static propTypes = {
+    ...ViewPropTypes,
+
+    /**
+     * Controls whether the view is a tab stop. Useful for buttons and other
+     * controls that can be focused.
+     *
+     * @platform windows
+     */
+    isTabStop: PropTypes.bool,
+
+    /**
+     * Sets the tab index for the view.
+     *
+     * @platform windows
+     */
+    tabIndex: PropTypes.number,
+
+    /**
+     * Sets the keyboard navigation mode for the view.
+     *
+     * @platform windows
+     */
+    tabNavigation: PropTypes.oneOf([
+      'local',
+      'cycle',
+      'once'
+    ]),
+
+    /**
+     * Controls whether control should use system default provided focus rects
+     *
+     * @platform windows
+     */
+    disableSystemFocusVisuals: PropTypes.bool,
+
+    /**
+     * Called when the view receives focus.
+     *
+     * @platform windows
+     */
+    onFocus: PropTypes.func,
+
+    /**
+     * Called when the view focus is lost.
+     *
+     * @platform windows
+     */
+    onBlur: PropTypes.func,
+
+    /**
+     * Set of keys that should be handled on key down by this component.
+     *
+     * @platform windows
+     */
+    handledKeyDownKeys: PropTypes.arrayOf(PropTypes.number),
+
+    /**
+     * Set of keys that should be handled on key up by this component.
+     *
+     * @platform windows
+     */
+    handledKeyUpKeys: PropTypes.arrayOf(PropTypes.number),
+
+    /**
+     * Called when key down while component has focus.
+     *
+     * @platform windows
+     */
+    onKeyDown: PropTypes.func,
+
+    /**
+     * Called when key up while component has focus.
+     *
+     * @platform windows
+     */
+    onKeyUp: PropTypes.func,
+  };
+
+  static keys = UIManager.WindowsControl.Constants.Keys;
+
+  /**
+   * Splits styles into two sets: one applying to focusable view, and the other one to the embedded child
+   * The only style we care about is "transform"
+   */
+  static splitStyle(styles) {
```

Comment:
>splitStyle [](start = 9, length = 10)

Have you considered solving this in native? For example:
  - We could introduce a new concept of "TransformDelegate". FocusableView can set itself to be the "TransformDelegate" of its child. When "transform" is set, it's set on the "TransformDelegate" if the node has one. Otherwise, it's set on the node itself. We can store "TransformDelegates" here: [https://github.com/Microsoft/react-native-windows/blob/4483d33ab2283abda24e981de22733f1627bfff8/ReactWindows/ReactNative.Shared/UIManager/DependencyObjectExtensions.cs](https://github.com/Microsoft/react-native-windows/blob/4483d33ab2283abda24e981de22733f1627bfff8/ReactWindows/ReactNative.Shared/UIManager/DependencyObjectExtensions.cs)
  - When AddChild is called on FocusableView, it can set "TransformDelegate" and transfer the transform from its child to itself in case somebody already set a transform before the parenting occurred.

Reply:
Yeah, I thought of doing it in native land, basically this way. 

It originally looked a little ugly to me (like "delegated layout" ugly, but I had no other choice for that one), so I hoped the JS solution would be better/shorter.
If the JS solution hadn't needed developers to use those helpers we wouldn't have had this discussion. But they do have to do that, so it's also ugly as is.
I'm torn. @rozele, should I attempt the native side of things?


---
In reply to: [157285100](https://github.com/Microsoft/react-native-windows/pull/1546#discussion_r157285100) [](ancestors = 157285100)

Reply:
Attempting a solution, may not be that ugly as I had thought.

---
In reply to: [157295016](https://github.com/Microsoft/react-native-windows/pull/1546#discussion_r157295016) [](ancestors = 157295016,157285100)